### ISSUE TITLE
Is there a way to detect if AI tracking is blocked by browser extension?

### ISSUE THREAD

**Issue by everscending** [microsoft/ApplicationInsights-JS#2011]
**Is your feature request related to a problem? Please describe.**

There is evidence that Application Insights tracking is being blocked by a 3rd party browser extension such as Ad Blocker or Privacy Badger.  Is there a method to detect when this is happening?

**Describe the solution you'd like**

When I reproduce the blocking with Privacy Badger I get the following in my JS console --

POST https://dc.services.visualstudio.com/v2/track net::ERR_BLOCKED_BY_CLIENT

I'd like to be able to implement a listener that will fire when this happens in order to give instructions to the user to disable their blocker.  This would be ideal to implement on the NotificationManager like so:

        const notificationManager = new NotificationManager();
        notificationManager.addNotificationListener({
          trackingBlocked: events => {
            console.log('Tracking is blocked by your browser or browser extension.  Events not tracked.. ', events);
          },
        });
        appInsights.loadAppInsights(false, undefined, notificationManager);

**Describe alternatives you've considered**

Intercepting calls to XMLHttpRequest.send.  The custom onerror handler is called but the event object does not contain any info about the error --

        const savedSend = XMLHttpRequest.prototype.send;
        XMLHttpRequest.prototype.send = function (data) {
          this.onerror = event => {
            console.log('ERROR!..', event, this);
          };
          savedSend.call(this, data);
        };

**Additional context**

**Comment by MSNev** [comment]
I know when we have tried this in the past (lot older browsers), JavaScript doesn't get enough information to make the determination that the request was "blocked" by the client. We just get a status code of zero, which is also returned for other use cases (not just the blocked case).

I've tagged as an enhancement (so it won't get auto closed) and we will schedule some time (not sure when at this point) to try and re-evaluate this. If you have cycles and would like to try and identify this your welcome to create a PR.

Please be aware that we are about to promote the beta release to master this week or next, so if you create a branch shortly merging could be a nightmare :-)

**Comment by everscending** [comment]
I tried testing the endpoint during the bootstrapping of the app with the following:

        const req = new XMLHttpRequest();
        req.addEventListener('load', () => {
          console.log('LOADED OK!');
        });

        req.addEventListener('error', event => {
          console.log('ERROR!', event);
        });

        req.open('POST', 'https://dc.services.visualstudio.com/v2/track');
        req.send();

This scenario is a slight improvement, I know the error is directly tied to call to the AI endpoint but it still doesn't indicate why it failed as you pointed out.  I wonder if that is even possible.  :\

**Comment by everscending** [comment]
On a related note, a possible solution to side-step this issue entirely is to proxy the /track calls through the backend server for the instrumented app --
https://github.com/shaneochotny/ApplicationInsights-JS-Proxy

**Comment by MSNev** [comment]
Yes, 

that's also possible and it would require that you also host the SDK (assuming your loading it from a CDN) on the same host. In the case listed it is using (by default) `https://www.shaneochotny.com/scripts/ai.0.js` which is a now very old (deprecated) version of the SDK and snippet.

But this too can get blocked (if you get enough attention to your site from people how do not want tracking data) where Ad-blockers support partial URL blocking not just full domain blocking (but it is less likely)

**Comment by mfreeman-xtivia** [comment]
Does anyone have a current working solution for this problem?

I have successfully proxied through my Node.JS server but I am losing the address information for the browser client in the process, i.e. the events show up in AppInsights with the IP/address of my server and not the original client

**Comment by MSNev** [comment]
If your proxying the request, then (in theory -- I've not looked at the backend code) you should be able to [add this header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For) to the proxied request and it ( should ) the converted

**Comment by mfreeman-xtivia** [comment]
Yeah that's what I thought too, but no matter of supplying that header, either in hardcoded form, or copying it from the client's initial request, seem to make any difference. The AppInsights record always seems to show the address information for the proxy and NOT the original client, which is what I am trying to make happen

**Comment by MSNev** [comment]
Hmm, just been going through an old copy of the server code that I have access to and it does look like that it checks for the following (note: it does validate that the IP address is correctly formed) in this order and will use the first one it finds
- x-forwarded-for
- forwarded-for
- x-forwarded
- forwarded
- x-client-ip
- client-ip
- x-real-ip
- real-ip

And if the header contains more than 1 valid value (some of the above headers support multiple comma separated values) it will choose the first one...

So maybe your server is also behind some proxy which is (potentially) append / changing the x-forwarded-for header???

**Comment by mfreeman-xtivia** [comment]
@MSNev Nev

By "server code" do you mean the server code for the actual AppInsights ingestor?

And if so, do you know if that logic somehow knows if its "browser telemetry" vs "server telemetry" and does anything different in those 2 cases?

**Comment by MSNev** [comment]
> By "server code" do you mean the server code for the actual AppInsights ingestor?

Yes

> By "server code" do you mean the server code for the actual AppInsights ingestor?

No, it appears to be just looking at the Http Request properties and processes the contents of the request (headers, body, etc). There is nothing obvious that I saw that seems to identify them separately, but I was just looking for the client ip processing and having worked on backends in a previous role (not for Application Insights) because HTTP is stateless there isn't much that can be safely assumed beyond the contents of the request.

**Comment by a-stankevich** [comment]
I have the same issue not being able to proxy Application Insights calls without losing Client IP information.

Has anyone found a solution for that?
